File: Projetos/ACBrLib/Demos/Java/Boleto/Imports/ACBrLibBoleto/src/com/acbr/boleto/ACBrBoleto.java
package com.acbr.boleto;

import com.acbr.ACBrLibBase;
import com.acbr.ACBrSessao;
import com.sun.jna.Library;
import com.sun.jna.Native;
import com.sun.jna.Platform;
import com.sun.jna.ptr.IntByReference;
import java.io.File;
import java.nio.ByteBuffer;
import java.nio.file.Paths;

public final class ACBrBoleto extends ACBrLibBase {   
    
    private interface ACBrBoletoLib extends Library {
        static String JNA_LIBRARY_NAME = LibraryLoader.getLibraryName();
        public final static ACBrBoletoLib INSTANCE = LibraryLoader.getInstance();       

        class LibraryLoader {
            private static String library = "";
            private static ACBrBoletoLib instance = null;

            public static String getLibraryName() {
                if (library.isEmpty()) {
                    if(Platform.isWindows()){
                        library = Platform.is64Bit() ? "ACBrBoleto64" : "ACBrBoleto32";                        
                    }else{
                        library = Platform.is64Bit() ? "acbrboleto64" : "acbrboleto32";
                    }
                }
                return library;
            }           

            public static ACBrBoletoLib getInstance() {
                if (instance == null) {
                    instance = (ACBrBoletoLib) Native.synchronizedLibrary((Library) Native.load(JNA_LIBRARY_NAME, ACBrBoletoLib.class));
                }
                return instance;
            }
        }

        int Boleto_Inicializar(String eArqConfig, String eChaveCrypt);

        int Boleto_Finalizar();

        int Boleto_Nome(ByteBuffer buffer, IntByReference bufferSize);

        int Boleto_Versao(ByteBuffer buffer, IntByReference bufferSize);

        int Boleto_UltimoRetorno(ByteBuffer buffer, IntByReference bufferSize);

        int Boleto_ConfigImportar(String eArqConfig);
        
        int Boleto_ConfigExportar(ByteBuffer buffer, IntByReference bufferSize);
        
        int Boleto_ConfigLer(String eArqConfig);

        int Boleto_ConfigGravar(String eArqConfig);

        int Boleto_ConfigLerValor(String eSessao, String eChave, ByteBuffer buffer, IntByReference bufferSize);

        int Boleto_ConfigGravarValor(String eSessao, String eChave, String valor);

        int Boleto_ConfigurarDados(String eArquivoIni);

        int Boleto_IncluirTitulos(String eArquivoIni, String eTpSaida);

        int Boleto_LimparLista(); 

        int Boleto_TotalTitulosLista();

        int Boleto_Imprimir(String eNomeImpressora);
        
        int Boleto_ImprimirBoleto(int eIndice, String eNomeImpressora);

        int Boleto_GerarPDF();

        int Boleto_GerarHTML();

        int Boleto_GerarRemessa(String eDir, int eNumArquivo, String eNomeArquivo);

        int Boleto_LerRetorno(String eDir, String eNomeArq);
        
        int Boleto_ObterRetorno(String eDir, String eNomeArq, ByteBuffer buffer, IntByReference bufferSize);

        int Boleto_EnviarEmail(String ePara, String eAssunto, String eMensagem, String eCC);

        int Boleto_EnviarEmailBoleto(int eIndice, String ePara, String eAssunto, String eMensagem, String eCC);
        
        int Boleto_SetDiretorioArquivo(String eDir, String eArq);

        int Boleto_ListaBancos(ByteBuffer buffer, IntByReference bufferSize);

        int Boleto_ListaCaractTitulo(ByteBuffer buffer, IntByReference bufferSize);

        int Boleto_ListaOcorrencias(ByteBuffer buffer, IntByReference bufferSize);

        int Boleto_ListaOcorrenciasEX(ByteBuffer buffer, IntByReference bufferSize);

        int Boleto_TamNossoNumero(String eCarteira, String enossoNumero, String eConvenio);

        int Boleto_CodigosMoraAceitos(ByteBuffer buffer, IntByReference bufferSize);

        int Boleto_SelecionaBanco(String eCodBanco);

        int Boleto_MontarNossoNumero(int eIndice, ByteBuffer buffer, IntByReference bufferSize);

        int Boleto_RetornaLinhaDigitavel(int eIndice, ByteBuffer buffer, IntByReference bufferSize);

        int Boleto_RetornaCodigoBarras(int eIndice, ByteBuffer buffer, IntByReference bufferSize);
        
        int Boleto_EnviarBoleto(int eCodigoOperacao, ByteBuffer buffer, IntByReference bufferSize);
        
        int Boleto_ConsultarTitulosPorPeriodo(String eArquivoIni, ByteBuffer buffer, IntByReference bufferSize);
    }
    
    public ACBrBoleto() throws Exception {
        File iniFile = Paths.get(System.getProperty("user.dir"), "ACBrLib.ini").toFile();
        if (!iniFile.exists()) {
            iniFile.createNewFile();
        }

        int ret = ACBrBoletoLib.INSTANCE.Boleto_Inicializar(toUTF8(iniFile.getAbsolutePath()), toUTF8(""));
        checkResult(ret);
    }

    public ACBrBoleto(String eArqConfig, String eChaveCrypt) throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_Inicializar(toUTF8(eArqConfig), toUTF8(eChaveCrypt));
        checkResult(ret);
    }
    
    @Override
    protected void dispose() throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_Finalizar();
        checkResult(ret);
    }
    
    public String nome() throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrBoletoLib.INSTANCE.Boleto_Nome(buffer, bufferLen);
        checkResult(ret);

        return processResult(buffer, bufferLen);
    }

    public String versao() throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrBoletoLib.INSTANCE.Boleto_Versao(buffer, bufferLen);
        checkResult(ret);

        return processResult(buffer, bufferLen);
    }
    
     public void configLer() throws Exception {
         configLer("");
     }
    
    public void configLer(String eArqConfig) throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_ConfigLer(toUTF8(eArqConfig));
        checkResult(ret);
    }
    
    public void configGravar() throws Exception {
        configGravar("");
    }
    
    public void configGravar(String eArqConfig) throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_ConfigGravar(toUTF8(eArqConfig));
        checkResult(ret);
    }

    @Override
    public String configLerValor(ACBrSessao eSessao, String eChave) throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrBoletoLib.INSTANCE.Boleto_ConfigLerValor(toUTF8(eSessao.name()), toUTF8(eChave), buffer, bufferLen);
        checkResult(ret);
        
        return processResult(buffer, bufferLen);
    }

    @Override
    public void configGravarValor(ACBrSessao eSessao, String eChave, Object value) throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_ConfigGravarValor(toUTF8(eSessao.name()), toUTF8(eChave), toUTF8(value.toString()));
        checkResult(ret);
    }
    
    public void importarConfig(String eArqConfig) throws Exception {        
        int ret = ACBrBoletoLib.INSTANCE.Boleto_ConfigImportar(eArqConfig);
        checkResult(ret);        
    }
    
    public String exportarConfig() throws Exception {		
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrBoletoLib.INSTANCE.Boleto_ConfigExportar(buffer, bufferLen);
        checkResult(ret);

        return fromUTF8(buffer, bufferLen.getValue());		
    }
    
    public void ConfigurarDados(String eArquivoIni) throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_ConfigurarDados(toUTF8(eArquivoIni));
        checkResult(ret);
    }
    
    public void IncluirTitulos(String eArquivoIni, String eTpSaida) throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_IncluirTitulos(toUTF8(eArquivoIni), toUTF8(eTpSaida));
        checkResult(ret);
    }
    
    public void LimparLista() throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_LimparLista();
        checkResult(ret);
    }
    
    public int TotalTitulosLista() throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_TotalTitulosLista();
        checkResult(ret);
        
        return ret;
    }
    
    public void Imprimir() throws Exception {
        Imprimir("");
    }
    
    public void Imprimir(String eNomeImpressora) throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_Imprimir(eNomeImpressora);
        checkResult(ret);
    }
    
    public void ImprimirBoleto(int Indice) throws Exception {
        ImprimirBoleto(Indice, "");
    }
    
    public void ImprimirBoleto(int Indice, String eNomeImpressora) throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_ImprimirBoleto(Indice, eNomeImpressora);
        checkResult(ret);
    }
    
    public void GerarPDF() throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_GerarPDF();
        checkResult(ret);
    }

    public void GerarHTML() throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_GerarHTML();
        checkResult(ret);
    }

    public void GerarRemessa(String eDir, int eNumArquivo, String eNomeArquivo) throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_GerarRemessa(eDir, eNumArquivo, eNomeArquivo);
        checkResult(ret);
    }
    
    public void LerRetorno(String eDir, String eNomeArq) throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_LerRetorno(eDir, eNomeArq);
        checkResult(ret);
    }

    public void EnviarEmail(String ePara, String eAssunto, String eMensagem, String eCC) throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_EnviarEmail(ePara, eAssunto, eMensagem, eCC);
        checkResult(ret);
    }
    
    public void EnviarEmailBoleto(int eIndice, String ePara, String eAssunto, String eMensagem, String eCC) throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_EnviarEmailBoleto(eIndice, ePara, eAssunto, eMensagem, eCC);
        checkResult(ret);
    }
    
    public void SetDiretorioArquivo(String eDir, String eArq) throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_SetDiretorioArquivo(eDir, eArq);
        checkResult(ret);
    }

    public String[] ListaBancos() throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrBoletoLib.INSTANCE.Boleto_ListaBancos(buffer, bufferLen);
        checkResult(ret);
        
        return processResult(buffer, bufferLen).split("|");
    }
    
    public String[] ListaCaractTitulo() throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrBoletoLib.INSTANCE.Boleto_ListaCaractTitulo(buffer, bufferLen);
        checkResult(ret);
        
        return processResult(buffer, bufferLen).split("|");
    }
    
    public String[] ListaOcorrencias() throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrBoletoLib.INSTANCE.Boleto_ListaOcorrencias(buffer, bufferLen);
        checkResult(ret);
        
        return processResult(buffer, bufferLen).split("|");
    }

    public String[] ListaOcorrenciasEX() throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrBoletoLib.INSTANCE.Boleto_ListaOcorrenciasEX(buffer, bufferLen);
        checkResult(ret);
        
        return processResult(buffer, bufferLen).split("|");
    }
    
    public int TamNossoNumero(String eCarteira, String enossoNumero, String eConvenio) throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_TamNossoNumero(eCarteira, enossoNumero, eConvenio);
        checkResult(ret);
        
        return ret;
    }
    
    public String CodigosMoraAceitos() throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrBoletoLib.INSTANCE.Boleto_CodigosMoraAceitos(buffer, bufferLen);
        checkResult(ret);
        
        return processResult(buffer, bufferLen);
    }
    
    public void SelecionaBanco(String eCodBanco) throws Exception {
        int ret = ACBrBoletoLib.INSTANCE.Boleto_SelecionaBanco(eCodBanco);
        checkResult(ret);
    }

    public String MontarNossoNumero(int eIndice) throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrBoletoLib.INSTANCE.Boleto_MontarNossoNumero(eIndice, buffer, bufferLen);
        checkResult(ret);
        
        return processResult(buffer, bufferLen);
    }

    public String RetornaLinhaDigitavel(int eIndice) throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrBoletoLib.INSTANCE.Boleto_RetornaLinhaDigitavel(eIndice, buffer, bufferLen);
        checkResult(ret);
        
        return processResult(buffer, bufferLen);
    }

    public String RetornaCodigoBarras(int eIndice) throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);

        int ret = ACBrBoletoLib.INSTANCE.Boleto_RetornaCodigoBarras(eIndice, buffer, bufferLen);
        checkResult(ret);
        
        return processResult(buffer, bufferLen);
    }

    public String EnviarBoleto(int eCodigoOperacao) throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);
        
        int ret = ACBrBoletoLib.INSTANCE.Boleto_EnviarBoleto(eCodigoOperacao, buffer, bufferLen);
        checkResult(ret);
        
        return processResult(buffer, bufferLen);
    }
    
    public String ConsultarTitulosPorPeriodo(String eArquivoIni) throws Exception {
        ByteBuffer buffer = ByteBuffer.allocate(STR_BUFFER_LEN);
        IntByReference bufferLen = new IntByReference(STR_BUFFER_LEN);
        
        int ret = ACBrBoletoLib.INSTANCE.Boleto_ConsultarTitulosPorPeriodo(eArquivoIni, buffer, bufferLen);
        checkResult(ret);
        
        return processResult(buffer, bufferLen);
    }
    
    @Override
    protected void UltimoRetorno(ByteBuffer buffer, IntByReference bufferLen) {
        ACBrBoletoLib.INSTANCE.Boleto_UltimoRetorno(buffer, bufferLen);
    }
}

File: Projetos/ACBrLib/Demos/C#/Boleto/Imports/Dinamico/ST/ACBrBoleto.cs
using System;
using System.IO;
using System.Text;
using ACBrLib.Core;
using ACBrLib.Core.Boleto;

namespace ACBrLib.Boleto
{
    public sealed partial class ACBrBoleto : ACBrLibHandle
    {
        #region Constructors

        public ACBrBoleto(string eArqConfig = "", string eChaveCrypt = "") : base(IsWindows ? "ACBrBoleto64.dll" : "libacbrboleto64.so",
                                                                                  IsWindows ? "ACBrBoleto32.dll" : "libacbrboleto32.so")
        {
            var inicializar = GetMethod<Boleto_Inicializar>();
            var ret = ExecuteMethod<int>(() => inicializar(ToUTF8(eArqConfig), ToUTF8(eChaveCrypt)));

            CheckResult(ret);

            Config = new ACBrBoletoConfig(this);
        }

        #endregion Constructors

        #region Properties

        public string Nome
        {
            get
            {
                var bufferLen = BUFFER_LEN;
                var buffer = new StringBuilder(bufferLen);

                var method = GetMethod<Boleto_Nome>();
                var ret = ExecuteMethod<int>(() => method(buffer, ref bufferLen));

                CheckResult(ret);

                return ProcessResult(buffer, bufferLen);
            }
        }

        public string Versao
        {
            get
            {
                var bufferLen = BUFFER_LEN;
                var buffer = new StringBuilder(bufferLen);

                var method = GetMethod<Boleto_Versao>();
                var ret = ExecuteMethod<int>(() => method(buffer, ref bufferLen));

                CheckResult(ret);

                return ProcessResult(buffer, bufferLen);
            }
        }

        public ACBrBoletoConfig Config { get; }

        #endregion Properties

        #region Methods

        #region Ini

        public override void ConfigGravar(string eArqConfig = "")
        {
            var gravarIni = GetMethod<Boleto_ConfigGravar>();
            var ret = ExecuteMethod<int>(() => gravarIni(ToUTF8(eArqConfig)));

            CheckResult(ret);
        }

        public override void ConfigLer(string eArqConfig = "")
        {
            var lerIni = GetMethod<Boleto_ConfigLer>();
            var ret = ExecuteMethod<int>(() => lerIni(ToUTF8(eArqConfig)));

            CheckResult(ret);
        }

        public override T ConfigLerValor<T>(ACBrSessao eSessao, string eChave)
        {
            var method = GetMethod<Boleto_ConfigLerValor>();

            var bufferLen = BUFFER_LEN;
            var pValue = new StringBuilder(bufferLen);
            var ret = ExecuteMethod<int>(() => method(ToUTF8(eSessao.ToString()), ToUTF8(eChave), pValue, ref bufferLen));
            CheckResult(ret);

            var value = ProcessResult(pValue, bufferLen);
            return ConvertValue<T>(value);
        }

        public override void ConfigGravarValor(ACBrSessao eSessao, string eChave, object value)
        {
            if (value == null) return;

            var method = GetMethod<Boleto_ConfigGravarValor>();
            var propValue = ConvertValue(value);

            var ret = ExecuteMethod<int>(() => method(ToUTF8(eSessao.ToString()), ToUTF8(eChave), ToUTF8(propValue)));
            CheckResult(ret);
        }

        public override void ImportarConfig(string eArqConfig = "")
        {
            var importarConfig = GetMethod<Boleto_ConfigImportar>();
            var ret = ExecuteMethod<int>(() => importarConfig(ToUTF8(eArqConfig)));

            CheckResult(ret);
        }

        public override string ExportarConfig()
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Boleto_ConfigExportar>();
            var ret = ExecuteMethod<int>(() => method(buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen);
        }

        #endregion Ini

        public void ConfigurarDados(params BoletoInfo[] infos)
        {
            var iniFile = new ACBrIniFile();
            foreach (var info in infos)
                info.WriteToIni(iniFile);

            ConfigurarDados(iniFile.ToString());
        }

        public void ConfigurarDados(string eArquivoIni)
        {
            var method = GetMethod<Boleto_ConfigurarDados>();
            var ret = ExecuteMethod<int>(() => method(ToUTF8(eArquivoIni)));

            CheckResult(ret);
        }

        public void IncluirTitulos(params Titulo[] titulos)
        {
            var iniFile = new ACBrIniFile();
            for (var i = 0; i < titulos.Length; i++)
            {
                titulos[i].Index = i+1;
                titulos[i].WriteToIni(iniFile);
            }

            IncluirTitulos(iniFile.ToString());
        }

        public void IncluirTitulos(BoletoTpSaida eTpSaida, params Titulo[] titulos)
        {
            var iniFile = new ACBrIniFile();
            for (var i = 0; i < titulos.Length; i++)
            {
                titulos[i].Index = i+1;
                titulos[i].WriteToIni(iniFile);
            }

            IncluirTitulos(iniFile.ToString(), eTpSaida);
        }

        public void IncluirTitulos(string eArquivoIni, BoletoTpSaida? eTpSaida = null)
        {
            var tpSaida = $"{(eTpSaida.HasValue ? (char)eTpSaida.Value : ' ')}";

            var method = GetMethod<Boleto_IncluirTitulos>();
            var ret = ExecuteMethod<int>(() => method(ToUTF8(eArquivoIni), ToUTF8(tpSaida)));

            CheckResult(ret);
        }

        public void LimparLista()
        {
            var method = GetMethod<Boleto_LimparLista>();
            var ret = ExecuteMethod<int>(() => method());
            CheckResult(ret);
        }

        public int TotalTitulosLista()
        {
            var method = GetMethod<Boleto_TotalTitulosLista>();
            var ret = ExecuteMethod<int>(() => method());

            CheckResult(ret);

            return ret;
        }

        public void Imprimir(string eNomeImpressora = "")
        {
            var method = GetMethod<Boleto_Imprimir>();
            var ret = ExecuteMethod(() => method(ToUTF8(eNomeImpressora)));
            CheckResult(ret);
        }

        public void Imprimir(int indice, string eNomeImpressora = "")
        {
            var method = GetMethod<Boleto_ImprimirBoleto>();
            var ret = ExecuteMethod(() => method(indice, ToUTF8(eNomeImpressora)));
            CheckResult(ret);
        }

        public void GerarPDF()
        {
            var method = GetMethod<Boleto_GerarPDF>();
            var ret = ExecuteMethod(() => method());
            CheckResult(ret);
        }

        public void GerarPDF(Stream aStream)
        {
            if (aStream == null) throw new ArgumentNullException(nameof(aStream));

            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Boleto_SalvarPDF>();
            var ret = ExecuteMethod(() => method(buffer, ref bufferLen));

            CheckResult(ret);

            var pdf = ProcessResult(buffer, bufferLen);
            Base64ToStream(pdf, aStream);
        }

        public void GerarPDF(int indice)
        {
            var method = GetMethod<Boleto_GerarPDFBoleto>();
            var ret = ExecuteMethod(() => method(indice));
            CheckResult(ret);
        }

        public void GerarPDF(int indice, Stream aStream)
        {
            if (aStream == null) throw new ArgumentNullException(nameof(aStream));

            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Boleto_SalvarPDFBoleto>();
            var ret = ExecuteMethod(() => method(indice, buffer, ref bufferLen));

            CheckResult(ret);

            var pdf = ProcessResult(buffer, bufferLen);
            Base64ToStream(pdf, aStream);
        }

        public void GerarHTML()
        {
            var method = GetMethod<Boleto_GerarHTML>();
            var ret = ExecuteMethod<int>(() => method());
            CheckResult(ret);
        }

        public void GerarRemessa(string eDir, int eNumArquivo, string eNomeArq)
        {
            var method = GetMethod<Boleto_GerarRemessa>();
            var ret = ExecuteMethod<int>(() => method(ToUTF8(eDir), eNumArquivo, ToUTF8(eNomeArq)));

            CheckResult(ret);
        }

        public void GerarRemessaStream(string eDir, int eNumArquivo, string eNomeArq, Stream aStream) 
        {
            if (aStream == null) throw new ArgumentNullException(nameof(aStream));

            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Boleto_GerarRemessaStream>();
            var ret = ExecuteMethod(() => method(eDir, eNumArquivo, eNomeArq, buffer, ref bufferLen));

            CheckResult(ret);

            var rem = ProcessResult(buffer, bufferLen);
            Base64ToStream(rem, aStream);
        }

        public RetornoBoleto ObterRetorno(string eDir, string eNomeArq)
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Boleto_ObterRetorno>();
            var ret = ExecuteMethod(() => method(ToUTF8(eDir), ToUTF8(eNomeArq), buffer, ref bufferLen));

            CheckResult(ret);

            return RetornoBoleto.LerRetorno(ProcessResult(buffer, bufferLen));
        }

        public void LerRetorno(string eDir, string eNomeArq)
        {
            var method = GetMethod<Boleto_LerRetorno>();
            var ret = ExecuteMethod<int>(() => method(ToUTF8(eDir), ToUTF8(eNomeArq)));

            CheckResult(ret);
        }

        public void EnviarEmail(string ePara, string eAssunto, string eMensagem, string eCC)
        {
            var method = GetMethod<Boleto_EnviarEmail>();
            var ret = ExecuteMethod<int>(() => method(ToUTF8(ePara), ToUTF8(eAssunto), ToUTF8(eMensagem), ToUTF8(eCC)));

            CheckResult(ret);
        }

        public void EnviarEmailBoleto(int indice, string ePara, string eAssunto, string eMensagem, string eCC)
        {
            var method = GetMethod<Boleto_EnviarEmailBoleto>();
            var ret = ExecuteMethod<int>(() => method(indice, ToUTF8(ePara), ToUTF8(eAssunto), ToUTF8(eMensagem), ToUTF8(eCC)));

            CheckResult(ret);
        }

        public void SetDiretorioArquivo(string eDir, string eArq = "")
        {
            var method = GetMethod<Boleto_SetDiretorioArquivo>();
            var ret = ExecuteMethod<int>(() => method(ToUTF8(eDir), ToUTF8(eArq)));

            CheckResult(ret);
        }

        public string[] ListaBancos()
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Boleto_ListaBancos>();
            var ret = ExecuteMethod<int>(() => method(buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen).Split('|');
        }

        public string[] ListaCaractTitulo()
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Boleto_ListaCaractTitulo>();
            var ret = ExecuteMethod<int>(() => method(buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen).Split('|');
        }

        public string[] ListaOcorrencias()
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Boleto_ListaOcorrencias>();
            var ret = ExecuteMethod<int>(() => method(buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen).Split('|');
        }

        public string[] ListaOcorrenciasEX()
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Boleto_ListaOcorrenciasEX>();
            var ret = ExecuteMethod<int>(() => method(buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen).Split('|');
        }

        public int TamNossoNumero(string eCarteira, string enossoNumero, string eConvenio)
        {
            var method = GetMethod<Boleto_TamNossoNumero>();
            var ret = ExecuteMethod<int>(() => method(ToUTF8(eCarteira), ToUTF8(enossoNumero), ToUTF8(eConvenio)));

            CheckResult(ret);

            return ret;
        }

        public string CodigosMoraAceitos()
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Boleto_CodigosMoraAceitos>();
            var ret = ExecuteMethod<int>(() => method(buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen);
        }

        public void SelecionaBanco(string eCodBanco)
        {
            var method = GetMethod<Boleto_SelecionaBanco>();
            var ret = ExecuteMethod<int>(() => method(ToUTF8(eCodBanco)));

            CheckResult(ret);
        }

        public string MontarNossoNumero(int eIndex)
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Boleto_MontarNossoNumero>();
            var ret = ExecuteMethod<int>(() => method(eIndex, buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen);
        }

        public string RetornaLinhaDigitavel(int eIndex)
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Boleto_RetornaLinhaDigitavel>();
            var ret = ExecuteMethod<int>(() => method(eIndex, buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen);
        }

        public string RetornaCodigoBarras(int eIndex)
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Boleto_RetornaCodigoBarras>();
            var ret = ExecuteMethod<int>(() => method(eIndex, buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen);
        }

        public RetornoWeb EnviarBoleto(OperacaoBoleto opercao)
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Boleto_EnviarBoleto>();
            var ret = ExecuteMethod(() => method((int)opercao, buffer, ref bufferLen));

            CheckResult(ret);

            return RetornoWeb.LerRetorno(ProcessResult(buffer, bufferLen));
        }

        public string ConsultarTitulosPorPeriodo(string eArquivoIni)
        {
            var bufferLen = BUFFER_LEN;
            var buffer = new StringBuilder(bufferLen);

            var method = GetMethod<Boleto_ConsultarTitulosPorPeriodo>();
            var ret = ExecuteMethod<int>(() => method(eArquivoIni, buffer, ref bufferLen));

            CheckResult(ret);

            return ProcessResult(buffer, bufferLen);
        }

        #region Private Methods

        protected override void FinalizeLib()
        {
            var finalizar = GetMethod<Boleto_Finalizar>();
            var ret = ExecuteMethod<int>(() => finalizar());
            CheckResult(ret);
        }

        protected override string GetUltimoRetorno(int iniBufferLen = 0)
        {
            var bufferLen = iniBufferLen < 1 ? BUFFER_LEN : iniBufferLen;
            var buffer = new StringBuilder(bufferLen);
            var ultimoRetorno = GetMethod<Boleto_UltimoRetorno>();

            if (iniBufferLen < 1)
            {
                ExecuteMethod<int>(() => ultimoRetorno(buffer, ref bufferLen));
                if (bufferLen <= BUFFER_LEN) return FromUTF8(buffer);

                buffer.Capacity = bufferLen;
            }

            ExecuteMethod<int>(() => ultimoRetorno(buffer, ref bufferLen));
            return FromUTF8(buffer);
        }

        #endregion Private Methods

        #endregion Methods
    }
}